lib/rbs/json_schema/generator.rb

```diff
@@ -0,0 +1,344 @@
+require "json"
+require "rbs/json_schema/validation_error"
+
+module RBS
+  module JSONSchema
+    class Generator
+      attr_reader :options
+      attr_reader :stdout
+      attr_reader :stderr
+      attr_reader :name
+      attr_reader :decls
+
+      # SCHEMA DEFINITIONS required for generating RBS from JSON schema
+      SCHEMA_DEFINITIONS = %w(definitions properties type enum const allOf oneOf items $ref $schema title description).freeze
+
+      # REGEXP to resolve declaration of $ref
+      REF_REGEX_1 = /\#\/definitions\/[a-z]\w*\b/ # matches current schema definitions => #/definitions/__any_type_alias_name__
+      # REGEXP to resolve declaration of $ref
+      REF_REGEX_2 = /[a-z]\w*\b\.json$/ # matches schema filenames => __any_type_alias_name__.json
+      # REGEXP to resolve declaration of $ref
+      REF_REGEX_3 = /[a-z]\w*\b\.json\#\/definitions\/[a-z]\w*\b/ # matches external schema definitions => __any_type_alias_name__.json/#/definitions/__any_type_alias_name__
+
+      # Alias declarations for easier referencing
+      Alias = RBS::AST::Declarations::Alias
+      TypeName = RBS::TypeName
+      Namespace = RBS::Namespace
+
+      def initialize(options:, stdout:, stderr:)
+        @options = options
+        @stdout = stdout
+        @stderr = stderr
+        @name = nil
+        @decls = []
+      end
+
+      # Returns type name with prefixes & appropriate namespace
+      def type_name(name, prefix: nil, absolute: nil)
+        TypeName.new(
+          name: prefix ? "#{@name}__#{name}".to_sym : name.to_sym,
+          namespace: absolute ? Namespace.root : Namespace.empty
+        )
+      end
+
+      # IMPORTANT: Function invoked to generate RBS from JSON schema
+      # Generates RBS from JSON schema after validating options & writes to file/STDOUT
+      def generate
+        # Validate options received from CLI
+        validate_options()
+
+        # Read filename & it's contents
+        read_file do |name, json|
+          begin
+            stdout.puts "\n======= Generating RBS for schema: #{name} =======\n\n"
+            # Obtain schema in RUBY's Hash data structure by parsing file content
+            schema = JSON.parse(json)
+            @name = name
+          rescue JSON::ParserError => e
+            # Print error message for invalid JSON content
+            raise ValidationError.new(message: "Invalid JSON content!\n#{e.full_message}")
+          end
+
+          unless (schema.is_a?(Hash) && (schema.key?("$schema") || schema.key?("$id")))
+            raise ValidationError.new(message: "Invalid JSON Schema: #{schema}")
+          end
+
+          # Parse & generate declarations from `definitions` attribute of the schema
```

Comment:
The problem is that there is no standard for the `definitions` key.

* https://json-schema.org/understanding-json-schema/structuring.html#ref
* https://json-schema.org/understanding-json-schema/structuring.html#id19

`$defs` would be more standard key...